### PR TITLE
8295017: Remove Windows specific workaround in JLI_Snprintf

### PR DESCRIPTION
**Pull request by TheShermanTanker**
The C99 snprintf is available with Visual Studio 2015 and above, alongside Windows 10 and the UCRT, and is no longer identical to the outdated Windows _snprintf. Since support for the Visual C++ 2017 compiler was removed a while ago, we can now safely remove the compatibility workaround on Windows and have JLI_Snprintf simply delegate to snprintf.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295017](https://bugs.openjdk.org/browse/JDK-8295017): Remove Windows specific workaround in JLI_Snprintf

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [9149aae1](https://git.openjdk.org/jdk/pull/10625/files/9149aae17f022008c0bfce15a19cb67077c8cd25)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10625/head:pull/10625` \
`$ git checkout pull/10625`

Update a local copy of the PR: \
`$ git checkout pull/10625` \
`$ git pull https://git.openjdk.org/jdk pull/10625/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10625`

View PR using the GUI difftool: \
`$ git pr show -t 10625`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10625.diff">https://git.openjdk.org/jdk/pull/10625.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/native/libjli/jli_util.h [hunk 1]
```diff
@@ -81,14 +81,23 @@ JLI_GetAppArgIndex();
...
 #define JLI_StrPBrk(p1, p2)     strpbrk((p1), (p2))
 
+/*
+ * Visual Studio 2015 and above, alongside Windows 10 and the UCRT
+ * now use the C99 snprintf, which is no longer identical to the
+ * outdated Windows _snprintf, so we can remove the compatibility
+ * workaround.
+ * https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference
+ * /snprintf-snprintf-snprintf-l-snwprintf-snwprintf-l?view=msvc-170
+ */
```
Comment (kimbarrett) [inline src/java.base/share/native/libjli/jli_util.h:91]: I don't think the comment about the *lack* of a workaround is needed, just adding clutter.  But this isn't code I have much involvement with.  Other than that, the change looks fine.

Comment (dholmes-ora) [reply src/java.base/share/native/libjli/jli_util.h:91]: I agree, we don't document the absence of a workaround.

**Review by dholmes-ora — COMMENTED** [review]
Looks good modulo the comment block.

**Review by kimbarrett — COMMENTED** [review]
Looks good.

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks.

**Comment by TheShermanTanker** [comment]
/integrate
@dholmes-ora could I trouble you for a sponsor? Thanks!

**Comment by mrserb** [comment]
@TheShermanTanker Working on a similar cleanup, and wonder if is it correct to assume that the "snprintf" adds "nul" even in case of error.
For example, this code was removed by this patch:

```
    if (rc < 0) {
        /* apply ansi semantics */
        buffer[size - 1] = '\0';
        return (int)size;
    } else if (rc == size) {
        /* force a null terminator */
        buffer[size - 1] = '\0';
    }
```

If the result of "snprintf" was negative we always set the '\0'. But what about default "snprintf"?

**Comment by TheShermanTanker** [comment]
@mrserb snprintf in the UCRT does indeed null terminate the buffer, conforming to C99. The relevant behaviour can be found documented under https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/snprintf-snprintf-snprintf-l-snwprintf-snwprintf-l?view=msvc-170#behavior-summary

**Comment by mrserb** [comment]
Thank you!

>If processing string specifier s, S, or Z, format specification processing stops, a NULL is placed at the beginning of the buffer.

I hope this is not an MS extension/implementaion details since I did not find this in any other places.

**Comment by dholmes-ora** [comment]
> I hope this is not an MS extension/implementation detail since I did not find this in any other places.

@mrserb this change was to a Windows specific file.

**Comment by mrserb** [comment]
>@mrserb this change was to a Windows specific file.

That change removed the windows specific version of the JLI_Snprintf, and now we use `#define JLI_Snprintf                    snprintf` on all platforms. And my question was about that "cross-platform" `snprintf`. As linked in the comment above on Windows it adds the null at the start of the buffer in case of error when a negative value is returned. But is that specified by the c99?

**Comment by dholmes-ora** [comment]
@mrserb we already had the define for non-Windows, this change just made it unconditional. This change only had an affect on Windows where we removed the custom `JLI_Snprintf` and started using `snprintf` the same as all other platforms.

**Comment by dholmes-ora** [comment]
That said C99 snprintf does not say anything about writing a NUL terminator in the case where nothing at all is written.

**Pull request closed without merging**